runtime/shared_common/shrinit.cpp

```diff
@@ -3177,6 +3182,47 @@ j9shr_isBCIEnabled(J9JavaVM *vm)
 	return (0 != (vm->sharedClassConfig->runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_BCI));
 }
 
+static UDATA
+romToRamGetRomAddress(void *item)
+{
+	RomToRamQueryEntry *queryEntry = (RomToRamQueryEntry *)item;
+	UDATA romAddress = (UDATA)(queryEntry->romClass);
+	if (J9_ARE_ALL_BITS_SET(romAddress, ROM_TO_RAM_QUERY_TAG)) {
+		return romAddress & ~ROM_TO_RAM_QUERY_TAG;
+	} else {
+		RomToRamEntry *entry = (RomToRamEntry *)item;
+		return (UDATA)(entry->ramClass->romClass);
+	}
+}
+
+/* THREADING: Must be protected by J9SharedClassConfig.romToRamHashTableMutex */
+static UDATA
+romToRamHashFn(void *item, void *userData)
+{
+	return romToRamGetRomAddress(item);
+}
+
+/* THREADING: Must be protected by J9SharedClassConfig.romToRamHashTableMutex */
+static UDATA
+romToRamEqualFn(void *left, void *right, void *userData)
+{
+	return (romToRamGetRomAddress(left) == romToRamGetRomAddress(right));
+}
+
+static void
+romToRamRemoveEntry(J9HookInterface **hookInterface, UDATA eventNum, void *voidData, void *userData)
+{
+	J9VMClassesUnloadEvent *data = (J9VMClassesUnloadEvent *)voidData;
+	J9SharedClassConfig *config = data->currentThread->javaVM->sharedClassConfig;
+	for (J9Class *ramClass = data->classesToUnload; ramClass; ramClass = ramClass->gcLink) {
+		RomToRamEntry entry;
+		entry.ramClass = ramClass;
+		omrthread_rwmutex_enter_write(config->romToRamHashTableMutex);
```

Comment:
The monitor enter/exit code can be moved outside of the for loop so that we don't have to repeatedly enter and exit.